src/SpecsParser.sol

```diff
@@ -238,34 +248,89 @@ library ERC4337SpecsParser {
-                if (!isAssociated) {
+        // Loop over the debug steps to validate the storage accesses
+        for (uint256 i; i < debugTrace.length; i++) {
+            // Get the current debug step
+            VmSafe.DebugStep memory currentStep = debugTrace[i];
+            // Skip if the opcode is not SLOAD or SSTORE, TLOAD or TSTORE
+            if (
+                currentStep.opcode != 0x54 // SLOAD
+                    && currentStep.opcode != 0x55 // SSTORE
+                    && currentStep.opcode != 0x5C // TLOAD
+                    && currentStep.opcode != 0x5D // TSTORE
+            ) {
+                continue;
+            }
+            // Init current access account
+            address currentAccessAccount = currentStep.contractAddr;
+            // Init current access slot
+            bytes32 currentSlot = bytes32(uint256(currentStep.stack[0]));
+
+            /// Access to associated storage of the account in an external (non-entity) contract
+            // is allowed if either
+            if (!isEntity(entities, currentAccessAccount)) {
+                // [STO-021] The account already exists.
+                bool accountAlreadyExists = entities.account.code.length != 0
+                    || (
+                        currentAccessAccount == userOpDetails.entryPoint
+                            && entities.account != address(0)
+                    );
+                // [STO-022] There is an initCode and the factory contract is staked.
+                bool isFactoryStaked = entities.isFactoryStaked;
+                if (
+                    !(
+                        isAssociatedStorage(currentSlot, currentAccessAccount, entities)
+                            && (accountAlreadyExists || isFactoryStaked)
+                    )
+                ) {
+                    revert InvalidStorageLocation(
+                        currentAccessAccount,
+                        getLabel(currentAccessAccount),
+                        currentSlot,
+                        currentStep.opcode == 0x55 || currentStep.opcode == 0x5D
+                    );
+                }
+            }
+            // If the entity (paymaster, factory) is staked, then it is also
+            // allowed:
+            else if (isEntityAndStaked(entities, currentAccessAccount)) {
+                // [STO-010] Access to the “account” storage is always allowed.
+                if (currentAccessAccount == entities.account) {
+                    continue;
+                }
+                // [STO-031] Access the entity’s own storage.
+                else if (currentAccessAccount == userOpDetails.sender) {
+                    continue;
+                }
+                // [STO-032] Read/Write Access to storage slots that are associated with the entity,
+                // in any non-entity contract.
+                else if (
+                    !isEntity(entities, currentAccessAccount)
```

Comment:
when would this be true? arent we in the `isEntityAndStaked(entities, currentAccessAccount)` condition here?

Reply:
the outer condition should actually just check if either entity is staked. (the wording is a bit werid on the ERC for this condition but I think this should be correct. fixed in [5165b53]

Reply:
reworked the logic in [2978e2d]